unit-testing/unit-testing.md

```diff
@@ -0,0 +1,169 @@
+
+Try to imagine you are making some small changes to your production code and when you execute your tests a whole bunch of #1 or #2 tests fail.
+Will you know which of your changes broke the tests?
+
+No imagine all of the failed tests have names like #3.
+Tests like these can actually give you an understanding of the problem before even having to look at the failure or inside the test.
+
+**Here some pointers for finding a good name:**
+
+*  Be as precise as possible about what aspect of the method under test you are validating.
+*  Include the expected result inside you name. I.e. `nullIsAnInvalidArgument()`, `deletingAnEntityAlsoRemovesItFromCache()` etc.
+*  Don't shy away from longer names. Brevity is not always the way to go.
+  
+**Example:**
+
+```java
+@RunWith(Enclosed.class)
+public class FooTest {
+ 
+    public static class SomeMethod extends FooTest {
+ 
+        @Test
+        public void canBeInvokedByAuthenticatedUser() {
+            ...
+        }
+ 
+        @Test(expected=AuthenticationException.class)
+        public void cannotBeInvokedByUnauthenticatedUser() {
+            ...
+        }
+ 
+    }
+ 
+}
+```
+
+## Scoping
+
+One test should always test only one thing! Deciding what "one thing" means is harder than it sounds...
+
+The test in the following example has a very ambiguous name. It has the name because it actually tests two aspects of persisting the entity and that name is the best fit for both:
+
+* It verifies the save operation on the entity manager
+* and the cleanup of the used entity cache.
+
+Verifying both of these aspects of "persisting" in a single test makes it test two separate things.
+
+```java
+@Test
+public void entityCanBePersisted() {
+ 
+    Entity mockEntity = mock(Entity.class)
+ 
+    cut.persist(mockEntity);
+ 
+    verify(entityManager).save(mockEntity);
+    verify(entityCache).remove(mockEntity);
+ 
+}
+```
+
+This test can very easily be split into two distinct tests, each verifying a single aspect of the greater whole.
+
+```java
+@Test
+public void persistingAnEntitySavesItToTheEntityManager() {
+    Entity mockEntity = mock(Entity.class)
+    cut.persist(mockEntity);
+    verify(entityManager).save(mockEntity);
+}
+ 
+@Test
+public void persistingAnEntityClearsItToFromTheCache() {
+    Entity mockEntity = mock(Entity.class)
+    cut.persist(mockEntity);
+    verify(entityCache).remove(mockEntity);
+}
+```
+
+In combination with very precise naming this split allows for an easier identification of the source of a failed test. Each of this test can only fail because of one single reason. (unexpected exceptions aside)
+
+In this example the decision to split the single test into two separate tests is rather easy to make. In most cases it isn't that easy to known when and where to split. Developers should try to identify different aspects of the method under test which could be validated separately. In case these aspects are not obvious, writing a single test is the safer bet.
```

Comment:
This is the difficult part because it depends so much on how well you express the domain in the programming language. I'm not happy with "writing a single test is the safer bet". It implies that splitting tests is not as important as the other points.